### PR TITLE
Makefile: pull versions from centralized location and automatic prerequisites generation

### PR DESCRIPTION
**Pull request by vspinu**
A set of 3, hopefully non-contentious, changes to Makefiles:

   - Pick ESS (and ESSR) version from ess.el from where package.el, melpa, pkg-build, debian-elpa etc. pick it
   - Simplify lisp/Makefile by automatically generating prerequisite files
   - Add ESSR target to build and commit new versions of ESSR

### CODE REVIEW

File: lisp/ess-roxy.el [hunk 1]
```diff
@@ -68,7 +68,7 @@
 (require 'outline)
 (eval-when-compile
   (require 'cl-lib))
-(require 'essddr "ess-rd.el")
```
Comment (jabranham) [inline lisp/ess-roxy.el]: This change has already been made in `next`

File: lisp/Makefile [hunk 2]
```diff
@@ -56,28 +52,31 @@ uninstall:
...
 
 distclean: clean
-# and potentially more
 
-### Targets below here are only for developers - and these must have perl
+

+### File Targets
+
+.dependencies: $(ELS)
```
Comment (jabranham) [inline lisp/Makefile]: Why is this necessary?

File: VERSION [hunk 3]
```diff
@@ -1 +1,2 @@
-18.10.2
+!!! Auto generated by make. Update version in ess.el instead !!!
```
Comment (jabranham) [inline VERSION]: We should just remove this version file. I think it's only needed by a couple doc files and we can update them with sed.

Comment (vspinu) [reply VERSION]: It's a prerequisite for a few targets and it's not easy to get rid of but could be removed from git repo and ignored in .gitignore.

File: lisp/Makefile [hunk 4]
```diff
@@ -56,28 +52,31 @@ uninstall:
...
+.el.elc:
+	$(COMPILE) $<
 
 ## Should happen before building ESS; definitely *NOT* after unpacking tarball :
 julia-mode.el:
 	test -f ../etc/.IS.RELEASE || $(DOWNLOAD) \
 	https://raw.githubusercontent.com/JuliaEditorSupport/julia-emacs/master/julia-mode.el > $@
 
-ess-julia.elc: julia-mode.elc
+ess-julia.elc: julia-mode.el julia-mode.elc
```
[the remark this answers is not in the document]
Comment (vspinu) [reply lisp/Makefile]: Right. The old one wasn't right, the new one is not right either :)

File: lisp/ess.el [hunk 5]
```diff
@@ -16,8 +16,9 @@
 ;; Maintainer: ESS Core Team <ESS-core@r-project.org>
 ;; Copyright (C) 1997-2018 ESS Core Team <ESS-core@r-project.org>
 ;; Created: 7 Jan 1994
-;; Version: 18.10.2
+;; Version: 18.10.3-snapshot
```
[the remark this answers is not in the document]
Comment (vspinu) [reply lisp/ess.el]: Snapshot means " the build is a snapshot of your code at a given time". It's how things are done in java and clojure which I am familiar with but if -git is more suggestive let's do -git.

**Comment by vspinu** [comment]
And as I said before, it's good to keep all metadata in one well formatted file like most of packages do. All build tools are able to pick it from ess.el because they auto-generate ess-pkg.el and that's good enough. For tar.gz for package-install-file we can generate ess-pkg.el from ess.el, that's an one-liner; there is no contradiction there, or do you see one? There is much more going on in the issue which you cited, and much more that you don't see in there. Would appreciate if could not bring it up again.

File: lisp/Makefile (hunk 2 above)
Comment (vspinu) [reply lisp/Makefile]: Auto-generates the pre-requisites for all our elc files. We don't need to hard-code them in our Makefile any longer.

File: lisp/ess-roxy.el (hunk 1 above)
Comment (vspinu) [reply lisp/ess-roxy.el]: Maybe this one should go on next actually, or after you merge next.

File: lisp/Makefile (hunk 2 above)
Comment (vspinu) [reply lisp/Makefile]: There is actually a more elegant way of doing that with .SECONDEXPANSION but it doesn't work on ancient make 3.8 on Travis' ubuntu trusty. So I gave up after a night of struggle.

**Comment by vspinu** [comment]
> this is not where package.el would get it.

Let me attempt to put a period in this pkg.el dispute. Package.el does not mandate to keep NAME-pkg.el inside a repo. Packaging is unrelated to the organization of our code base. Its requirement is exclusive to the package archive which is defined as 

```
a multi-file package is stored in a package
archive as a tar file.  The tar file must be named ‘NAME-VERSION.tar’,
where NAME is the package name and VERSION is the version number.  
...
‘NAME-pkg.el’  ... must contain a single Lisp form, consisting of a call
to the function ‘define-package’,

```
The `define-package` form has only 3 arguments name, version, docstring and requirements. This means that we cannot add our other metadata there. So we will still have to keep other metadata in ess.el. Plus all lisp version vars (`ess-eversion`, `essr-version`, `ess-version` function etc) and probably much more metadata related tooling will be in the ess.el in the future. So it just doesn't make sense in my mind to keep this artificial split as part of the code base. All package build tools out there (package.el is not a build tool) auto-generate pkg.el and I think this is the right approach. So can we, if we really want to bypass build tools.  It's really common sense, not my subjective quirk.

File: lisp/ess.el (hunk 5 above)
[the remark this answers is not in the document]
Comment (jabranham) [reply lisp/ess.el]: Snapshot to me implies an intended release, sort of like a release candidate. I don't have a strong feeling about it though. `version-regexp-alist` treats `-snapshot` and `-git` the same I think.

File: VERSION (hunk 3 above)
Comment (vspinu) [reply VERSION]: Changelog, dist, essr and one RPM related target depend on VERSION file changes. We need a marker of some kind to keep track of the old version. I don't think there is a more elegant solution to this, but I am not a make expert.

File: lisp/Makefile (hunk 2 above)
Comment (jabranham) [reply lisp/Makefile]: Sorry, I wasn't clear. I was thinking more along the lines of "why is it necessary to specify .elc dependencies in the first place?" If .elc doesn't exist yet, the byte compiler will just use the corresponding .el file. Can't we just say "all .elc files depend on the corresponding .el file" and not care about the order?

Comment (vspinu) [reply lisp/Makefile]: But if one .el changes you need to recompile all els which depend on it. Or you wonder something else?

Comment (jabranham) [reply lisp/Makefile]: Ah, I knew I was missing something obvious. I wasn't thinking about non-clean checkouts.

File: lisp/ess-roxy.el [hunk 6]
```diff
@@ -938,7 +938,7 @@ passes ARGS to it."
            (ess-roxy-goto-end-of-roxy-comment)
            (looking-at "#")))
     (ess-roxy-with-filling-context t
-        (apply orig-fun args)))
+                                   (apply orig-fun args)))
```
Comment (jabranham) [inline lisp/ess-roxy.el]: This whitespace doesn't look right.

**Comment by vspinu** [comment]
I am pulling this in. I changed the ground truth of version back to VERSION file. There are too many uses of it in texi files and sed is not the right way to approach it there.
